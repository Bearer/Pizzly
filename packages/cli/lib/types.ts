export interface GlobalOptions {
    secretKey?: string;
    host?: string;
}

export type ENV = 'local' | 'staging' | 'production';

export interface DeployOptions extends GlobalOptions {
    env?: 'local' | 'staging' | 'production';
    local?: boolean;
    staging?: boolean;
    version?: string;
    sync?: string;
}
