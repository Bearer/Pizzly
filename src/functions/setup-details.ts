import Joi from '@hapi/joi'
import { Response, NextFunction } from 'express'
import { v4 as uuidv4 } from 'uuid'

import { TBackendRequestV4 } from '../types'
// import { getSetupDetails, saveSetupDetails } from '../clients/integrations'
import { asyncMiddleware } from '../errorHandler'
import { SetupDetailsNotFound } from '../errors'

const makeResponse = (payload: any) => ({
  StatusCode: 200,
  Payload: JSON.stringify(payload)
})

const reportError = (req: TBackendRequestV4, next: NextFunction, error: any) => {
  req.bearerResponse = makeResponse({ error })
  return next()
}

const getOptionalSetupDetails = async (req: TBackendRequestV4, setupId: string) => {
  const {
    // buid,
    // clientId,
    // stageVariables: { scopedUserDataTableId: scopedUserDataTableName }
  } = req

  try {
    // return await getSetupDetails({ buid, clientId, scopedUserDataTableName, setupId })
  } catch (e) {
    if (e instanceof SetupDetailsNotFound) {
      return undefined
    }

    throw e
  }
}

export const setupSave = asyncMiddleware(async (req: TBackendRequestV4, _res: Response, next: NextFunction) => {
  const { referenceId } = req.query
  const { setup } = req.body

  if (!setup) {
    return reportError(req, next, {
      code: 'MISSING_SETUP',
      message: 'Please provide a setup object containing information { setup: { ... } }'
    })
  }

  const { authType } = await req.integration.config()

  if (setup.type) {
    setup.type = setup.type.toUpperCase()
  } else {
    setup.type = authType
  }

  const { error: typeError } = Joi.validate(setup, formatValidator, { abortEarly: false })
  if (typeError) {
    return reportError(req, next, typeError)
  }

  if (setup.type !== authType) {
    return reportError(req, next, {
      code: 'INCONSISTENT_AUTH_TYPE',
      message:
        'The auth type of the supplied setup details must match the auth type of the API. ' +
        ` The supplied setup type was '${setup.type}' but the API is '${authType}'.`
    })
  }

  const { error } = Joi.validate(setup, validator[setup.type], { abortEarly: false })
  if (error) {
    return reportError(req, next, error)
  }

  // We can't store empty values
  for (const key in setup) {
    if (!setup[key]) {
      delete setup[key]
    }
  }

  const ref = referenceId || uuidv4()
  const exist = await getOptionalSetupDetails(req, ref)
  if (exist) {
    return reportError(req, next, {
      code: 'EXISTING_SETUP',
      message: 'A setup already exists with this reference'
    })
  }

  // await saveSetupDetails({ buid, clientId, scopedUserDataTableName, setupId: ref, data: setup })
  req.bearerResponse = makeResponse({ data: { ...setup, setupId: ref }, referenceId: ref })

  next()
})

export const setupRetrieve = asyncMiddleware(async (req: TBackendRequestV4, _res: Response, next: NextFunction) => {
  const { referenceId } = req.query

  if (!referenceId) {
    return reportError(req, next, {
      code: 'MISSING_PARAMETER',
      message: 'Please provide a referenceId'
    })
  }

  const data = await getOptionalSetupDetails(req, referenceId)
  req.bearerResponse = makeResponse({ referenceId, data: Boolean(data) })

  next()
})

enum AuthType {
  OAuth2 = 'OAUTH2',
  OAuth1 = 'OAUTH1'
}

const joiExtended = Joi.extend({
  name: 'authType',
  base: Joi.string()
    .valid(Object.values(AuthType))
    .required()
})

const oauth2Validator = Joi.object().keys({
  clientID: Joi.string()
    .required()
    .allow(''),
  clientSecret: Joi.string()
    .required()
    .allow(''),
  type: joiExtended.authType()
})

const oauth1Validator = Joi.object().keys({
  consumerKey: Joi.string()
    .required()
    .allow(''),
  consumerSecret: Joi.string()
    .required()
    .allow(''),
  type: joiExtended.authType()
})

const formatValidator = Joi.object()
  .keys({
    type: joiExtended.authType()
  })
  .unknown(true)

const validator = {
  [AuthType.OAuth2]: oauth2Validator,
  [AuthType.OAuth1]: oauth1Validator
}
