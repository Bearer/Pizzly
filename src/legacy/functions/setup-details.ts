import Joi from '@hapi/joi'
import { Response, NextFunction } from 'express'
import { v4 as uuidv4 } from 'uuid'

import { TBackendRequestV4 } from '../../types'
import { getSetupDetails, saveSetupDetails } from '../auth/clients/integrations'
import { asyncMiddleware } from '../../errorHandler'

export const setupSave = asyncMiddleware(async (req: TBackendRequestV4, res: Response, next: NextFunction) => {
  const { referenceId } = req.query
  const { buid, store } = req
  const { setup } = req.body

  console.log('[setupSave] body', req.body)
  console.log('[setupSave] setup', setup)
  console.log('[setupSave] buid', buid)

  if (!setup) {
    return res.send({
      code: 'MISSING_SETUP',
      message: 'Please provide a setup object containing information { setup: { ... } }'
    })
  }

  const { scopes } = setup
  console.log('[setupSave] scopes', scopes)

  const { error: typeError } = Joi.validate(setup, formatValidator, { abortEarly: false })
  if (typeError) {
    console.log('[setupSave] typeError', typeError)
    return res.send(typeError)
  }

  const { error } = Joi.validate(setup, validator[setup.type], { abortEarly: false })
  if (error) {
    console.log('[setupSave] error', error)
    return res.send(error)
  }

  // We can't store empty values
  for (const key in setup) {
    if (!setup[key]) {
      delete setup[key]
    }
  }

  const ref = referenceId || uuidv4()

  console.log('[setupSave] params', { buid, scopes, ref })
  await saveSetupDetails({ store, buid, setup, scopes, setupId: ref })

  res.send({
    referenceId: ref,
    ack: 'true'
  })
})

export const setupRetrieve = asyncMiddleware(async (req: TBackendRequestV4, res: Response, next: NextFunction) => {
  const { referenceId } = req.query
  const { store, buid } = req

  if (!referenceId) {
    return res.send({
      code: 'MISSING_PARAMETER',
      message: 'Please provide a referenceId'
    })
  }

  const data = await getSetupDetails({ store, buid, setupId: referenceId })
  res.send({
    referenceId,
    data: Boolean(data)
  })
})

enum AuthType {
  OAuth2 = 'OAUTH2',
  OAuth1 = 'OAUTH1'
}

const joiExtended = Joi.extend({
  name: 'authType',
  base: Joi.string()
    .valid(Object.values(AuthType))
    .required()
})

const oauth2Validator = Joi.object().keys({
  clientId: Joi.string()
    .required()
    .allow(''),
  clientSecret: Joi.string()
    .required()
    .allow(''),
  type: joiExtended.authType(),
  scopes: [Joi.array(), null]
})

const oauth1Validator = Joi.object().keys({
  consumerKey: Joi.string()
    .required()
    .allow(''),
  consumerSecret: Joi.string()
    .required()
    .allow(''),
  type: joiExtended.authType(),
  scopes: [Joi.array(), null]
})

const formatValidator = Joi.object()
  .keys({
    type: joiExtended.authType()
  })
  .unknown(true)

const validator = {
  [AuthType.OAuth2]: oauth2Validator,
  [AuthType.OAuth1]: oauth1Validator
}
