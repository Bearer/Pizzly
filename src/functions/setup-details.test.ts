import uuidv4 from 'uuid/v4'
import { MiddlewareTestHarness } from '../../tests/utils'
import { setupRetrieve, setupSave } from './setup-details'
import { TBackendRequestV4 } from '../types'
// import { getSetupDetails, saveSetupDetails } from '../clients/integrations'
import { SetupDetailsNotFound } from '../errors'

jest.mock('../clients/integrations')
jest.mock('uuid/v4')

// const setupDetails = { username: 'test-user' }
const clientId = 'test-client-id'
const buid = 'test-buid'
const setupId = 'test-setup-id'
const queryString = `?referenceId=${setupId}`

const noDetailsError = new SetupDetailsNotFound({ clientId, buid, setupId })

beforeEach(() => {
  // @ts-ignore
  // getSetupDetails.mockClear()
  // @ts-ignore
  // saveSetupDetails.mockClear()
})

describe('setupSave', () => {
  const generatedSetupId = 'test-generated-id'
  const invalidAuthTypeBody = { setup: { type: 'unknown' } }
  const body = { setup: { type: 'basic', username: 'test-user' } }

  beforeEach(() => {
    // @ts-ignore
    uuidv4.mockReturnValue(generatedSetupId)
  })

  const setup = (authType = 'BASIC') =>
    new MiddlewareTestHarness<TBackendRequestV4>({
      testMiddleware: setupSave,
      configureRequest: req => {
        req.clientId = clientId
        req.buid = buid
        req.integration = {
          config: () => Promise.resolve({ authType })
        }
      }
    })

  describe('when no setup parameter is given', () => {
    it('returns a MISSING_SETUP error', async () => {
      const test = setup()

      await test.get().expect(200)

      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })

  describe('when the auth type is unknown', () => {
    it('returns a validation error', async () => {
      const test = setup()

      await test
        .get()
        .send(invalidAuthTypeBody)
        .expect(200)

      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })

  describe('when the auth type does NOT match the integration config', () => {
    it('returns an error', async () => {
      const test = setup('OAUTH2')

      await test
        .get()
        .send(body)
        .expect(200)

      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })

  describe('when the setup details already exist', () => {
    it('returns an EXISTING_SETUP error', async () => {
      const test = setup()

      await test
        .get()
        .send(body)
        .expect(200)

      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })

  describe('when the setup details do NOT already exist', () => {
    beforeEach(() => {
      // @ts-ignore
      getSetupDetails.mockRejectedValueOnce(noDetailsError)
    })

    it('saves the setup details', async () => {
      const test = setup()

      await test
        .get()
        .send(body)
        .expect(200)

      // expect(getSetupDetails).toMatchSnapshot()
      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })

  describe('when a referenceId is provided', () => {
    beforeEach(() => {
      // @ts-ignore
      getSetupDetails.mockRejectedValueOnce(noDetailsError)
    })

    it('uses the provided id', async () => {
      const test = setup()

      await test
        .get(queryString)
        .send(body)
        .expect(200)

      // expect(getSetupDetails).toMatchSnapshot()
      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })

  describe('validation', () => {
    describe.each([
      [
        'OAUTH2',
        [
          { clientID: 'aClientId', clientSecret: 'aSecret', type: 'OAUTH2' },
          { clientID: 'aClientId', clientSecret: 'aSecret', type: 'OaUtH2' },
          { clientID: '', clientSecret: '', type: 'OAUTH2' }
        ],
        [
          { type: 'OAUTH2' },
          { clientId: 'aClientId', clientSecret: 'a secret', type: 'OAUTH2' },
          { clientID: 'a clientId', clientSecre: 'a secret', type: 'OAUTH2' },
          { clientID: 1, clientSecret: 4, type: 'OAUTH2' }
        ]
      ],
      [
        'APIKEY',
        [{ apiKey: 'anap asd asd qw3489012737ASDASDASDikey', type: 'APIKEY' }, { apiKey: '', type: 'APIKEY' }],
        [{ type: 'APIKEY' }, { apikey: 'anapikey', type: 'APIKEY' }]
      ],
      [
        'BASIC',
        [
          { username: '12737(*^&$@!)(*&#^(%!ikey 47986 key', password: 'asdasd', type: 'BASIC' },
          { username: '', password: '', type: 'BASIC' },
          { username: 'onlyMe', type: 'BASIC' }
        ],
        [{ type: 'BASIC' }, { Ysername: 'anapikey', passwordd: 'asdasd', type: 'BASIC' }]
      ],
      [
        'OAUTH1',
        [
          { consumerKey: 'asdasd', consumerSecret: 'asdasd=09823bas!d9123', type: 'OAUTH1' },
          { consumerKey: '', consumerSecret: '', type: 'OAUTH1' }
        ],
        [{ type: 'OAUTH1' }, { ConsumerKey: 'qwerty', consumer_secret: '1234testing=12', type: 'OAUTH1' }]
      ]
    ])('for auth type %s', (type: string, successes: any[], failures: any[]) => {
      describe.each(failures)('with input %j', failure => {
        it('fails', async () => {
          const test = setup(type)

          await test
            .get()
            .send({ setup: failure })
            .expect(200)

          expect(test.req.bearerResponse).toMatchSnapshot()
        })
      })

      describe.each(successes)('with input %j', success => {
        it('succeeds', async () => {
          const test = setup(type)

          await test
            .get()
            .send({ setup: success })
            .expect(200)

          expect(test.req.bearerResponse).toMatchSnapshot()
        })
      })
    })
  })

  describe('when empty strings are provided', () => {
    it('saves Basic auth setup details', async () => {
      const body = { setup: { type: 'BASIC', username: '', password: '' } }

      const test = setup()

      await test
        .get(queryString)
        .send(body)
        .expect(200)

      // expect(getSetupDetails).toMatchSnapshot()
      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })
})

describe('setupRetrieve', () => {
  const setup = () =>
    new MiddlewareTestHarness<TBackendRequestV4>({
      testMiddleware: setupRetrieve
    })

  describe('when no referenceId is given', () => {
    it('returns a MISSING_PARAMETER error', async () => {
      const test = setup()

      await test.get().expect(200)

      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })

  describe('when the setup details exist', () => {
    it('returns a response with data true', async () => {
      const test = setup()

      await test.get(queryString).expect(200)

      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })

  describe('when the setup details do NOT exist', () => {
    it('returns a response with data false', async () => {
      const test = setup()

      await test.get(queryString).expect(200)

      expect(test.req.bearerResponse).toMatchSnapshot()
    })
  })
})
