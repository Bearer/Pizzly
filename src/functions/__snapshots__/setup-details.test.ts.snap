// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupRetrieve when no referenceId is given returns a MISSING_PARAMETER error 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"code\\":\\"MISSING_PARAMETER\\",\\"message\\":\\"Please provide a referenceId\\"}}",
  "StatusCode": 200,
}
`;

exports[`setupRetrieve when the setup details do NOT exist returns a response with data false 1`] = `
Object {
  "Payload": "{\\"referenceId\\":\\"test-setup-id\\",\\"data\\":false}",
  "StatusCode": 200,
}
`;

exports[`setupRetrieve when the setup details exist returns a response with data true 1`] = `
Object {
  "Payload": "{\\"referenceId\\":\\"test-setup-id\\",\\"data\\":false}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type APIKEY with input {"apiKey":"","type":"APIKEY"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"APIKEY\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type APIKEY with input {"apiKey":"anap asd asd qw3489012737ASDASDASDikey","type":"APIKEY"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"apiKey\\":\\"anap asd asd qw3489012737ASDASDASDikey\\",\\"type\\":\\"APIKEY\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type APIKEY with input {"apikey":"anapikey","type":"APIKEY"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"apiKey\\\\\\" is required\\",\\"path\\":[\\"apiKey\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"apiKey\\",\\"label\\":\\"apiKey\\"}},{\\"message\\":\\"\\\\\\"apikey\\\\\\" is not allowed\\",\\"path\\":[\\"apikey\\"],\\"type\\":\\"object.allowUnknown\\",\\"context\\":{\\"child\\":\\"apikey\\",\\"value\\":\\"anapikey\\",\\"key\\":\\"apikey\\",\\"label\\":\\"apikey\\"}}],\\"_object\\":{\\"apikey\\":\\"anapikey\\",\\"type\\":\\"APIKEY\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type APIKEY with input {"type":"APIKEY"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"apiKey\\\\\\" is required\\",\\"path\\":[\\"apiKey\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"apiKey\\",\\"label\\":\\"apiKey\\"}}],\\"_object\\":{\\"type\\":\\"APIKEY\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type BASIC with input {"Ysername":"anapikey","passwordd":"asdasd","type":"BASIC"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"username\\\\\\" is required\\",\\"path\\":[\\"username\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"username\\",\\"label\\":\\"username\\"}},{\\"message\\":\\"\\\\\\"Ysername\\\\\\" is not allowed\\",\\"path\\":[\\"Ysername\\"],\\"type\\":\\"object.allowUnknown\\",\\"context\\":{\\"child\\":\\"Ysername\\",\\"value\\":\\"anapikey\\",\\"key\\":\\"Ysername\\",\\"label\\":\\"Ysername\\"}},{\\"message\\":\\"\\\\\\"passwordd\\\\\\" is not allowed\\",\\"path\\":[\\"passwordd\\"],\\"type\\":\\"object.allowUnknown\\",\\"context\\":{\\"child\\":\\"passwordd\\",\\"value\\":\\"asdasd\\",\\"key\\":\\"passwordd\\",\\"label\\":\\"passwordd\\"}}],\\"_object\\":{\\"Ysername\\":\\"anapikey\\",\\"passwordd\\":\\"asdasd\\",\\"type\\":\\"BASIC\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type BASIC with input {"type":"BASIC"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"username\\\\\\" is required\\",\\"path\\":[\\"username\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"username\\",\\"label\\":\\"username\\"}}],\\"_object\\":{\\"type\\":\\"BASIC\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type BASIC with input {"username":"","password":"","type":"BASIC"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"BASIC\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type BASIC with input {"username":"12737(*^&$@!)(*&#^(%!ikey 47986 key","password":"asdasd","type":"BASIC"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"username\\":\\"12737(*^&$@!)(*&#^(%!ikey 47986 key\\",\\"password\\":\\"asdasd\\",\\"type\\":\\"BASIC\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type BASIC with input {"username":"onlyMe","type":"BASIC"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"username\\":\\"onlyMe\\",\\"type\\":\\"BASIC\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH1 with input {"ConsumerKey":"qwerty","consumer_secret":"1234testing=12","type":"OAUTH1"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"consumerKey\\\\\\" is required\\",\\"path\\":[\\"consumerKey\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"consumerKey\\",\\"label\\":\\"consumerKey\\"}},{\\"message\\":\\"\\\\\\"consumerSecret\\\\\\" is required\\",\\"path\\":[\\"consumerSecret\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"consumerSecret\\",\\"label\\":\\"consumerSecret\\"}},{\\"message\\":\\"\\\\\\"ConsumerKey\\\\\\" is not allowed\\",\\"path\\":[\\"ConsumerKey\\"],\\"type\\":\\"object.allowUnknown\\",\\"context\\":{\\"child\\":\\"ConsumerKey\\",\\"value\\":\\"qwerty\\",\\"key\\":\\"ConsumerKey\\",\\"label\\":\\"ConsumerKey\\"}},{\\"message\\":\\"\\\\\\"consumer_secret\\\\\\" is not allowed\\",\\"path\\":[\\"consumer_secret\\"],\\"type\\":\\"object.allowUnknown\\",\\"context\\":{\\"child\\":\\"consumer_secret\\",\\"value\\":\\"1234testing=12\\",\\"key\\":\\"consumer_secret\\",\\"label\\":\\"consumer_secret\\"}}],\\"_object\\":{\\"ConsumerKey\\":\\"qwerty\\",\\"consumer_secret\\":\\"1234testing=12\\",\\"type\\":\\"OAUTH1\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH1 with input {"consumerKey":"","consumerSecret":"","type":"OAUTH1"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"OAUTH1\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH1 with input {"consumerKey":"asdasd","consumerSecret":"asdasd=09823bas!d9123","type":"OAUTH1"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"consumerKey\\":\\"asdasd\\",\\"consumerSecret\\":\\"asdasd=09823bas!d9123\\",\\"type\\":\\"OAUTH1\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH1 with input {"type":"OAUTH1"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"consumerKey\\\\\\" is required\\",\\"path\\":[\\"consumerKey\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"consumerKey\\",\\"label\\":\\"consumerKey\\"}},{\\"message\\":\\"\\\\\\"consumerSecret\\\\\\" is required\\",\\"path\\":[\\"consumerSecret\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"consumerSecret\\",\\"label\\":\\"consumerSecret\\"}}],\\"_object\\":{\\"type\\":\\"OAUTH1\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH2 with input {"clientID":"","clientSecret":"","type":"OAUTH2"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"OAUTH2\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH2 with input {"clientID":"a clientId","clientSecre":"a secret","type":"OAUTH2"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"clientSecret\\\\\\" is required\\",\\"path\\":[\\"clientSecret\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"clientSecret\\",\\"label\\":\\"clientSecret\\"}},{\\"message\\":\\"\\\\\\"clientSecre\\\\\\" is not allowed\\",\\"path\\":[\\"clientSecre\\"],\\"type\\":\\"object.allowUnknown\\",\\"context\\":{\\"child\\":\\"clientSecre\\",\\"value\\":\\"a secret\\",\\"key\\":\\"clientSecre\\",\\"label\\":\\"clientSecre\\"}}],\\"_object\\":{\\"clientID\\":\\"a clientId\\",\\"clientSecre\\":\\"a secret\\",\\"type\\":\\"OAUTH2\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH2 with input {"clientID":"aClientId","clientSecret":"aSecret","type":"OAUTH2"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"clientID\\":\\"aClientId\\",\\"clientSecret\\":\\"aSecret\\",\\"type\\":\\"OAUTH2\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH2 with input {"clientID":"aClientId","clientSecret":"aSecret","type":"OaUtH2"} succeeds 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"clientID\\":\\"aClientId\\",\\"clientSecret\\":\\"aSecret\\",\\"type\\":\\"OAUTH2\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH2 with input {"clientID":1,"clientSecret":4,"type":"OAUTH2"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"clientID\\\\\\" must be a string\\",\\"path\\":[\\"clientID\\"],\\"type\\":\\"string.base\\",\\"context\\":{\\"value\\":1,\\"key\\":\\"clientID\\",\\"label\\":\\"clientID\\"}},{\\"message\\":\\"\\\\\\"clientSecret\\\\\\" must be a string\\",\\"path\\":[\\"clientSecret\\"],\\"type\\":\\"string.base\\",\\"context\\":{\\"value\\":4,\\"key\\":\\"clientSecret\\",\\"label\\":\\"clientSecret\\"}}],\\"_object\\":{\\"clientID\\":1,\\"clientSecret\\":4,\\"type\\":\\"OAUTH2\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH2 with input {"clientId":"aClientId","clientSecret":"a secret","type":"OAUTH2"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"clientID\\\\\\" is required\\",\\"path\\":[\\"clientID\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"clientID\\",\\"label\\":\\"clientID\\"}},{\\"message\\":\\"\\\\\\"clientId\\\\\\" is not allowed\\",\\"path\\":[\\"clientId\\"],\\"type\\":\\"object.allowUnknown\\",\\"context\\":{\\"child\\":\\"clientId\\",\\"value\\":\\"aClientId\\",\\"key\\":\\"clientId\\",\\"label\\":\\"clientId\\"}}],\\"_object\\":{\\"clientId\\":\\"aClientId\\",\\"clientSecret\\":\\"a secret\\",\\"type\\":\\"OAUTH2\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave validation for auth type OAUTH2 with input {"type":"OAUTH2"} fails 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"clientID\\\\\\" is required\\",\\"path\\":[\\"clientID\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"clientID\\",\\"label\\":\\"clientID\\"}},{\\"message\\":\\"\\\\\\"clientSecret\\\\\\" is required\\",\\"path\\":[\\"clientSecret\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"clientSecret\\",\\"label\\":\\"clientSecret\\"}}],\\"_object\\":{\\"type\\":\\"OAUTH2\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave when a referenceId is provided uses the provided id 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"BASIC\\",\\"username\\":\\"test-user\\",\\"setupId\\":\\"test-setup-id\\"},\\"referenceId\\":\\"test-setup-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave when a referenceId is provided uses the provided id 2`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"BASIC\\",\\"username\\":\\"test-user\\",\\"setupId\\":\\"test-setup-id\\"},\\"referenceId\\":\\"test-setup-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave when empty strings are provided saves Basic auth setup details 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"BASIC\\",\\"setupId\\":\\"test-setup-id\\"},\\"referenceId\\":\\"test-setup-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave when no setup parameter is given returns a MISSING_SETUP error 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"code\\":\\"MISSING_SETUP\\",\\"message\\":\\"Please provide a setup object containing information { setup: { ... } }\\"}}",
  "StatusCode": 200,
}
`;

exports[`setupSave when the auth type does NOT match the integration config returns an error 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"code\\":\\"INCONSISTENT_AUTH_TYPE\\",\\"message\\":\\"The auth type of the supplied setup details must match the auth type of the API.  The supplied setup type was 'BASIC' but the API is 'OAUTH2'.\\"}}",
  "StatusCode": 200,
}
`;

exports[`setupSave when the auth type is unknown returns a validation error 1`] = `
Object {
  "Payload": "{\\"error\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"type\\\\\\" must be one of [OAUTH2, OAUTH1, APIKEY, BASIC]\\",\\"path\\":[\\"type\\"],\\"type\\":\\"any.allowOnly\\",\\"context\\":{\\"value\\":\\"UNKNOWN\\",\\"valids\\":[\\"OAUTH2\\",\\"OAUTH1\\",\\"APIKEY\\",\\"BASIC\\"],\\"key\\":\\"type\\",\\"label\\":\\"type\\"}}],\\"_object\\":{\\"type\\":\\"UNKNOWN\\"}}}",
  "StatusCode": 200,
}
`;

exports[`setupSave when the setup details already exist returns an EXISTING_SETUP error 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"BASIC\\",\\"username\\":\\"test-user\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave when the setup details do NOT already exist saves the setup details 1`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"BASIC\\",\\"username\\":\\"test-user\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;

exports[`setupSave when the setup details do NOT already exist saves the setup details 2`] = `
Object {
  "Payload": "{\\"data\\":{\\"type\\":\\"BASIC\\",\\"username\\":\\"test-user\\",\\"setupId\\":\\"test-generated-id\\"},\\"referenceId\\":\\"test-generated-id\\"}",
  "StatusCode": 200,
}
`;
